### PR TITLE
[Feature](update) Support `update on current_timestamp`

### PR DESCRIPTION
**Pull request by bobhan1**
## Proposed changes

If a column is declared with `on update current_timestamp`, its value will be updated to the current timestamp whenever there is an update on the row.

Example:
```
mysql> CREATE TABLE t1 (
    ->     k int,
    ->     v int,
    ->     update_time datetime default current_timestamp on update current_timestamp)
    -> UNIQUE KEY(k) DISTRIBUTED BY HASH(k) BUCKETS 1
    -> PROPERTIES(
    ->     "replication_num" = "1",
    ->     "enable_unique_key_merge_on_write" = "true"
    -> );
Query OK, 0 rows affected (0.01 sec)
mysql> set enable_unique_key_partial_update=true;
Query OK, 0 rows affected (0.00 sec)

mysql> set enable_insert_strict=false;
Query OK, 0 rows affected (0.00 sec)

mysql> insert into t1(k,v) values(1,1),(2,2);
Query OK, 2 rows affected (0.04 sec)
mysql> select * from t1;
+------+------+---------------------+
| k    | v    | update_time         |
+------+------+---------------------+
|    1 |    1 | 2023-10-25 22:38:10 |
|    2 |    2 | 2023-10-25 22:38:10 |
+------+------+---------------------+
2 rows in set (0.03 sec)
mysql> set enable_unique_key_partial_update=false;
Query OK, 0 rows affected (0.00 sec)

mysql> insert into t1(k,v) values(1,100);
Query OK, 1 row affected (0.04 sec)
mysql> select * from t1;
+------+------+---------------------+
| k    | v    | update_time         |
+------+------+---------------------+
|    2 |    2 | 2023-10-25 22:38:10 |
|    1 |  100 | 2023-10-25 22:38:37 |
+------+------+---------------------+
2 rows in set (0.03 sec)
```

similar to https://dev.mysql.com/doc/refman/8.0/en/timestamp-initialization.html

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

File: docs/zh-CN/docs/sql-manual/sql-reference/Data-Definition-Statements/Create/CREATE-TABLE.md [hunk 1]
```diff
@@ -129,6 +129,10 @@ distribution_desc
             // 只用于DATETIME类型，导入数据缺失该值时系统将赋予当前时间
             dt DATETIME DEFAULT CURRENT_TIMESTAMP
         ```
+  * `on update current_timestamp`
```
Comment (zhannngchen) [inline docs/zh-CN/docs/sql-manual/sql-reference/Data-Definition-Statements/Create/CREATE-TABLE.md:132]: should support unique key table merge-on-write implementation only

File: regression-test/suites/unique_with_mow_p0/partial_update/test_update_on_current_timestamp.groovy [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+// under the License.
+
+suite("test_mow_update_on_current_timestamp", "p0") {
+    sql 'set experimental_enable_nereids_planner=true'
+    sql 'set enable_fallback_to_original_planner=false'
+    sql 'set enable_nereids_dml=true'
+    sql "sync;"
+
+
+    def t1 = "nereids_update_on_current_timestamp1"
```
Comment (zhannngchen) [inline regression-test/suites/unique_with_mow_p0/partial_update/test_update_on_current_timestamp.groovy:26]: Add cases that using `on update current_timestamp` for other data model.

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

File: regression-test/suites/nereids_p0/insert_into_table/update_on_current_timestamp.groovy [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+                    `dft` int(11) DEFAULT "4321",
+                    `update_time` datetime(6) default current_timestamp(4) on update current_timestamp(3))
+                UNIQUE KEY(`id`) DISTRIBUTED BY HASH(`id`) BUCKETS 1
+                PROPERTIES(
+                    "replication_num" = "1",
+                    "enable_unique_key_merge_on_write" = "true");"""
+        exception "The precision of the default value of column[update_time] should be the same with the precision in 'ON UPDATE CURRENT_TIMESTAMP'."
+    }
+
+    // illegal case 4: use 'update on current_timestamp' on
```
Comment (zhannngchen) [inline regression-test/suites/nereids_p0/insert_into_table/update_on_current_timestamp.groovy:134]: not complete comment？

**Comment by bobhan1** [comment]
run buildall

**Review by zhannngchen — DISMISSED** [review]
LGTM

File: regression-test/suites/nereids_p0/insert_into_table/update_on_current_timestamp.groovy [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    qt_sql "select count(distinct update_time2) from ${t1} where update_time2 > '2023-10-01 00:00:00';"
+
+
+    // set enable_unique_key_partial_update=true, it's a partial update
+    // don't specify the `update_time` column
+    // it will be automatically updated to current_timestamp()
+    sql "set enable_unique_key_partial_update=true;"
+    sql "set enable_insert_strict=false;"
+    sql "sync;"
+    sql """ insert into ${t1}(id, score) values
```
Comment (zhannngchen) [inline regression-test/suites/nereids_p0/insert_into_table/update_on_current_timestamp.groovy:62]: also add a case for `update` statement

**Comment by bobhan1** [comment]
run buildall

**Review by zhannngchen — DISMISSED** [review]
LGTM

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/info/ColumnDefinition.java [hunk 5]
```diff
@@ -198,6 +221,33 @@ public void validate(Set<String> keysSet, boolean isEnableMergeOnWrite, KeysType
...
             }
         }
+        if (onUpdateDefaultValue.isPresent()
+                && onUpdateDefaultValue.get().getValue() != null
+                && type.toCatalogDataType().isScalarType()) {
+            try {
+                ColumnDef.validateDefaultValue(type.toCatalogDataType(),
+                        onUpdateDefaultValue.get().getValue(), onUpdateDefaultValue.get().getDefaultValueExprDef());
+            } catch (Exception e) {
+                throw new AnalysisException(e.getMessage(), e);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/info/ColumnDefinition.java:231]: ditto

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run p0

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by bobhan1** [comment]
run feut

**Comment by bobhan1** [comment]
run pipelinex_p0

**Comment by bobhan1** [comment]
run feut

**Comment by bobhan1** [comment]
run feut

**Comment by bobhan1** [comment]
run feut

**Review by zhannngchen — APPROVED** [review]
LGTM

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**